rts/motoko-rts/src/region.rs

```diff
@@ -491,54 +470,59 @@ pub unsafe fn region_new<M: Memory>(mem: &mut M) -> Value {
         }
     };
 
-    let next_id = meta_data::total_allocated_regions::get() as u16;
+    let next_id = meta_data::total_allocated_regions::get();
 
     if next_id == meta_data::max::REGIONS {
         region_trap_with("out of regions")
     };
 
-    meta_data::total_allocated_regions::set(next_id as u64 + 1);
+    meta_data::total_allocated_regions::set(next_id + 1);
 
     let vec_pages = alloc_blob(mem, Bytes(0));
     allocation_barrier(vec_pages);
-    let r_ptr = alloc_region(mem, next_id as u32, 0, vec_pages);
+    let r_ptr = alloc_region(mem, next_id, 0, vec_pages);
 
-    // Update Region table.
-    {
-        let r_id = RegionId::from_id(next_id);
-        let c = meta_data::region_table::get(&r_id);
-        // sanity check: Region table says that this region is available.
-        assert_eq!(c, None);
-        meta_data::region_table::set(&r_id, Some(RegionSizeInPages(0)));
-    }
     r_ptr
 }
 
 pub unsafe fn region_recover<M: Memory>(mem: &mut M, rid: &RegionId) -> Value {
     use meta_data::bytes_of;
     use meta_data::size::PAGES_IN_BLOCK;
 
-    let c = meta_data::region_table::get(&rid);
-    let page_count = match c {
-        Some(RegionSizeInPages(pc)) => pc,
-        _ => {
-            rts_trap_with("region_recover failed: Expected Some(RegionSizeInPages(_))");
-        }
+    if rid.0 >= meta_data::total_allocated_regions::get() {
+        region_trap_with("cannot recover un-allocated region");
     };
 
-    debug_assert!(page_count < (u32::MAX - (PAGES_IN_BLOCK - 1)) as u64);
+    let tb = meta_data::total_allocated_blocks::get();
+
+    // determine page_count of this region
+    let mut page_count: u32 = 0;
+    {
+        let mut block_id: u16 = 0;
+        while (block_id as u32) < tb {
+            match meta_data::block_region_table::get(BlockId(block_id)) {
+                None => {}
+                Some((rid_, _rank, block_page_count)) => {
+                    if &rid_ == rid {
+                        page_count += block_page_count as u32;
+                    }
+                }
+            }
+            block_id += 1;
+        }
```

Comment:
A matter of taste, maybe a for loop would be simpler:

```suggestion
        for block_id in 0..tb as u16 {
            match meta_data::block_region_table::get(BlockId(block_id)) {
                None => {}
                Some((rid_, _rank, block_page_count)) => {
                    if &rid_ == rid {
                        page_count += block_page_count as u32;
                    }
                }
            }
        }
```